trl/trainer/orpo_trainer.py

```diff
@@ -0,0 +1,945 @@
+    import deepspeed
+
+
+class ORPOTrainer(Trainer):
+    r"""
+    Initialize ORPOTrainer.
+
+    Args:
+        model (`transformers.PreTrainedModel`):
+            The model to train, preferably an `AutoModelForSequenceClassification`.
+        args (`ORPOConfig`):
+            The ORPO config arguments to use for training.
+        data_collator (`transformers.DataCollator`):
+            The data collator to use for training. If None is specified, the default data collator (`DPODataCollatorWithPadding`) will be used
+            which will pad the sequences to the maximum length of the sequences in the batch, given a dataset of paired sequences.
+        train_dataset (`datasets.Dataset`):
+            The dataset to use for training.
+        eval_dataset (`datasets.Dataset`):
+            The dataset to use for evaluation.
+        tokenizer (`transformers.PreTrainedTokenizerBase`):
+            The tokenizer to use for training. This argument is required if you want to use the default data collator.
+        model_init (`Callable[[], transformers.PreTrainedModel]`):
+            The model initializer to use for training. If None is specified, the default model initializer will be used.
+        callbacks (`List[transformers.TrainerCallback]`):
+            The callbacks to use for training.
+        optimizers (`Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR]`):
+            The optimizer and scheduler to use for training.
+        preprocess_logits_for_metrics (`Callable[[torch.Tensor, torch.Tensor], torch.Tensor]`):
+            The function to use to preprocess the logits before computing the metrics.
+        peft_config (`Dict`, defaults to `None`):
+            The PEFT configuration to use for training. If you pass a PEFT configuration, the model will be wrapped in a PEFT model.
+        compute_metrics (`Callable[[EvalPrediction], Dict]`, *optional*):
+            The function to use to compute the metrics. Must take a `EvalPrediction` and return
+            a dictionary string to metric values.
+    """
+
+    _tag_names = ["trl", "orpo"]
+
+    def __init__(
+        self,
+        model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
+        args: Optional[ORPOConfig] = None,
+        data_collator: Optional[DataCollator] = None,
+        train_dataset: Optional[Dataset] = None,
+        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        model_init: Optional[Callable[[], PreTrainedModel]] = None,
+        callbacks: Optional[List[TrainerCallback]] = None,
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+        peft_config: Optional[Dict] = None,
+        compute_metrics: Optional[Callable[[EvalLoopOutput], Dict]] = None,
+    ):
```

Comment:
Wouldn't it be nice to also include here the `ORPO` specific args from the `ORPOConfig` so that users can still provide `TrainingArguments` to `args` within the `ORPOTrainer` while still being able to provide i.e. `beta`? Asking because that implementation would be closer to the current `DPOTrainer` and friends cc @lewtun

Reply:
Yes, it works, but those are not exposed as part of the `__init__` of the `ORPOTrainer` as in `DPOTrainer`, see https://github.com/huggingface/trl/blob/423991c204c4b36d4aea6955eeb53140fbbea9bb/trl/trainer/dpo_trainer.py#L144-L179

Reply:
Hey @alvarobartt as the SFT and DPO trainers have evolved, it has become clear that it's better to group all the non-callable arguments into a single config because otherwise you end up with a situation where some parameters are set in the `args` and others are set in the `__init__` which is confusing + harder to maintain.

We've started applying this approach to new trainers like KTO and ORPO and will gradually refactor the others to follow this. Is there some cases where it would be beneficial to preserve the current `__init__` based approach to set certain hyperparameters?

Reply:
Oh not at all, it was more in terms of alignment, as it was easier to always assume that one had to instantiate the `TrainingArguments` and then the additional stuff on top of the `...Trainer.__init__`, because otherwise one needs to import the specific configuration dataclass i.e. `ORPOConfig`, but I guess it's just to educate users to use the current approach to avoid `__init__` arg-duplication and to use the `...Config` dataclasses instead of just `TrainingArguments` 👍🏻

Reply:
Anyway, even if keeping only `args` there's still an issue with the values for both `max_length` and `max_prompt_length` that when provided the `ORPOTrainer.__init__` breaks, I'll leave the solution unresolved and resolve the rest of the comments, thanks!